unblob/handlers/archive/_safe_tarfile.py

```diff
@@ -0,0 +1,101 @@
+import os
+from pathlib import Path
+from tarfile import SUPPORTED_TYPES, ExtractError, TarFile
+
+from structlog import get_logger
+
+from unblob.extractor import is_safe_path
+
+logger = get_logger()
+
+
+class SafeTarFile(TarFile):
+    def extract(  # noqa: C901
+        self, member, path="", set_attrs=True, *, numeric_owner=False
+    ):
+        """Extract a member from the archive to the current working directory,
+        using its full name. Its file information is extracted as accurately
+        as possible. `member' may be a filename or a TarInfo object. You can
+        specify a different directory using `path'. File attributes (owner,
+        mtime, mode) are set unless `set_attrs' is False. If `numeric_owner`
+        is True, only the numbers for user/group names are used and not
+        the names.
+        """
+        self._check("r")  # type: ignore
+
+        path = Path(str(path))
+
+        if isinstance(member, str):
+            tarinfo = self.getmember(member)
+        else:
+            tarinfo = member
+
+        # Prepare the link target for makelink().
+        if tarinfo.islnk():
+            # MOD: traversal check
+            linkname = Path(tarinfo.linkname)
+            if is_safe_path(path, linkname):
+                tarinfo._link_target = path.joinpath(linkname).as_posix()  # type: ignore
+            else:
+                logger.warn("traversal attempt", path=tarinfo.linkname)
+
+        try:
+            # MOD: traversal check
+            name = Path(str(tarinfo.name))
+            if is_safe_path(path, name):
+                self._extract_member(
+                    tarinfo,
+                    path.joinpath(name).resolve().as_posix(),
+                    set_attrs=set_attrs,
+                    numeric_owner=numeric_owner,
+                )
+            else:
+                logger.warn("traversal attempt", path=tarinfo.name)
+        except OSError:
+            if self.errorlevel is not None and self.errorlevel > 0:
+                raise
+        except ExtractError:
+            if self.errorlevel is not None and self.errorlevel > 1:
+                raise
+
+    def _extract_member(  # noqa: C901
+        self, tarinfo, targetpath, set_attrs=True, numeric_owner=False
+    ):
```

Comment:
It may be a bit more tricky but the same applies here. The changes can be put in the overriden `extract` method (or checked post-fact in an exception handler for `PermissonError`.

Also, it'd make sense to cache if we are root, to spare a syscall for each special file.